File: LeetCode/652. 寻找重复的子树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        result = []
        subtrees = {}
        def subtree(node: Optional[TreeNode]) -> List[Optional[TreeNode]]:
            if not node:
                return '#'
            left = subtree(node.left)
            right = subtree(node.right)
            val = node.val
            subtreeStr = left + ',' + right + ',' + str(val)
            if subtreeStr not in subtrees:
                subtrees[subtreeStr] = 1
            elif subtrees[subtreeStr] == 1:
                result.append(node)
                subtrees[subtreeStr] += 1
            else:
                subtrees[subtreeStr] += 1
            return subtreeStr
        subtree(root)
        return result

File: LeetCode/652. 寻找重复的子树.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;
impl Solution {
    pub fn find_duplicate_subtrees(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Option<Rc<RefCell<TreeNode>>>> {
        let mut result = vec![];
        let mut subtrees = HashMap::new();
        Solution::subtree(root, &mut result, &mut subtrees);
        result
    }
    fn subtree(node: Option<Rc<RefCell<TreeNode>>>, mut result: &mut Vec<Option<Rc<RefCell<TreeNode>>>>, mut subtrees: &mut HashMap<String, i32>) -> String {
        match node {
            None => '#'.to_string(),
            Some(node) => {
                let val = node.borrow().val;
                let left = Solution::subtree(node.borrow().left.clone(), &mut result, &mut subtrees);
                let right = Solution::subtree(node.borrow().right.clone(), &mut result, &mut subtrees);

                let subtree = format!("{},{},{}", left, right, val);
                if let Some(count) = subtrees.get_mut(&subtree) {
                    if *count == 1 {
                        result.push(Some(node.clone()));
                    }
                    *count += 1;
                } else {
                    subtrees.insert(subtree.clone(), 1);
                }
                subtree
            }
        }
    }
}
